Handle CR without NL printed in EditorConsole

matthijskooijman opened the pull request:
When uploading a sketch using dfu-util (e.g. used by various STM32 cores), progress bar lines would be printed. However, each update would be printed on a new line, rather than overwriting the same line as intended. Even more, the console would not scroll when subsequent lines were added, which made it quite annoying to see the upload progress.

It turns out dfu-util uses lone `\r` to overwrite the previous line. However, `EditorConsole` would simply replace those with `\n` instead. But because `ConsoleOutputStream` only scrolled on `\n`, these extra lines would fall out of view on the bottom of the console.

I originally tried updating the scroll behaviour to include these lines, but then thought it would be even nicer to just properly handle `\r` characters instead. That took a bit more time than I had anticipated, but the result is fairly clean code, I think :-)

With this PR applied, the dfu-util now looks like this:

![Peek 2020-03-28 15-14](https://user-images.githubusercontent.com/194491/77827701-7f2c0800-7117-11ea-939f-a7135fcdc779.gif)

Without this PR, it was like this (note that most of the action happens *outside* of view below, until the upload is complete):

![Peek 2020-03-28 17-13](https://user-images.githubusercontent.com/194491/77827702-7fc49e80-7117-11ea-8569-5ebd4b06a049.gif)

As an additional test for partial line overwriting, I added this bit of shell script to a script that was used during upload:

```
echo -ne 'ABCDEF\r'
sleep 3;
echo -ne 'DEF\r'
sleep 3;
echo -ne 'X\n'
sleep 3;
echo -ne 'ZZZ\n'
```
Which works as expected:

![Peek 2020-03-28 15-13](https://user-images.githubusercontent.com/194491/77827700-7f2c0800-7117-11ea-9feb-956fd83c4603.gif)

Again, this would be a good candidate for some testcases. @cmaglie, do you have some notes you could share about how to run the testsuite in a meaningful way (especially just run a single testcase)?

facchinm commented:
LOVE. THIS. PR

vdeconinck commented on app/src/processing/app/EditorConsole.java:
```diff
@@ -179,10 +185,48 @@ public boolean isEmpty() {
+    // This regex is written so it will necessarily match any string
+    // completely if applied repeatedly. This is important because any
+    // part not matched would be silently dropped.
+    Matcher m = Pattern.compile("([^\r\n]*)([\r\n]*\n)?(\r+)?").matcher(str);
```
Knowing this will be called for each and every line printed to the console, I wonder if it wouldn't make sense to move the time consuming compiled pattern to a static variable at the top of the file:
`private static Pattern newLinePattern = Pattern.compile("([^\r\n]*)([\r\n]*\n)?(\r+)?");`
Then here in the insertString method, just call 
`Matcher m = newLinePattern.matcher(str);`
Maybe premature optimization is the root of all evil, but repeatedly compiling the same Pattern can be quite expensive...

matthijskooijman replied:
Good idea, I'll see about that soon.

matthijskooijman replied:
Done, just pushed this.

cmaglie commented:
Nice! Definitely I want to see this feature implemented!

> As an additional test for partial line overwriting, I added this bit of shell script to a script that was used during upload:

I'd like to see also a test when multiple lines are sent together like:

```
echo -ne 'ABCDEF\rDEF\rX\nZZZ'
sleep 3;
echo -ne 'ZZZ\rYYY\n'
```

> Again, this would be a good candidate for some testcases. @cmaglie, do you have some notes you could share about how to run the testsuite in a meaningful way (especially just run a single testcase)?

You can't run a single test from `ant` with the current `build.xml` AFAIK, I usually run them one by one from eclipse.

matthijskooijman commented:
> I'd like to see also a test when multiple lines are sent together like:

Good point. I might have already tested this, but let's do this regardless.

> You can't run a single test from ant with the current build.xml AFAIK, I usually run them one by one from eclipse.

Ah, I don't have eclipse set up anymore, so that's not an easy way out here :-)

matthijskooijman commented:
Ok, just pushed an update. I rebased on top of master and added a fixup commit with some improvements and a testcase. Took me a bit longer than expected after I started on this two days ago, but I produced a few more PRs with test-suite fixes and repo cleanup along the way, so that's nice :-p

As far as I'm concerned, I think this is now ready to merge (after squashing the fixup commit, of course).

matthijskooijman commented:
Just amended some more style fixes into the fixup commit and prepended some more commits to `EditorConsole` to make the testcases work (I had those already, but accidentally split them into the wrong pull request :-p). So *now* it is ready for merging :-)

facchinm commented:
Rebased and merged :tada:
